[OffloadBundler] Rework the ctor of `OffloadTargetInfo` to support AMDGPU's generic target

shiltian opened the pull request:
The current parsing logic for the target string assumes it follows the format `<kind>-<triple>-<target id>:<feature>`, such as `hipv4-amdgcn-amd-amdhsa-gfx1030:+xnack`.
Specifically, it assumes that `<target id>` does not contain any `-`, relying on `rsplit` for parsing.
However, this assumption breaks for AMDGPU's generic targets, which may contain one or more `-`, such as `gfx10-3-generic` or `gfx12-generic`.
As a result, the existing approach using `rstrip` is no longer reliable.

This patch reworks the parsing logic to handle target strings more robustly, including support for generic targets.
The bundler now strictly requires a 4-field target triple.
Additionally, a new Python helper function has been added to `config.py` to normalize the target triple into the 4-field format when it is not, ensuring tests pass reliably.

shiltian commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/122629?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#122629** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/122629?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈 <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/122629?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#122627** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/122627?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.
<!-- Current dependencies on/for this PR: -->

shiltian commented on clang/lib/Driver/OffloadBundler.cpp:
```diff
@@ -84,31 +84,19 @@ OffloadTargetInfo::OffloadTargetInfo(const StringRef Target,
-    this->TargetID = "";
-  }
+  // <kind>-<triple>[-<target id>]
+  // <tiple> := <arch>-<vendor>-<os>-<env>
```
found a typo: triple

shiltian commented:
I suppose this change is quite intrusive for any external use of `clang-offload-bundler`. It's gonna require all the targets listed in `-targets` have to be in the normalized form (4-field).

yxsamliu commented:
Did you try this patch with internal PSDB? Does it break any existing HIP apps or libraries?

shiltian commented:
> Did you try this patch with internal PSDB? Does it break any existing HIP apps or libraries?

Will make a PR shortly after the migration.

lamb-j commented:
Is there a way for us to still support 3-field triples and generic targets?

In the past we worked to allow 3-field triples as bundler inputs. I would be surprised if adding a 4-field restriction didn't break something internally.

shiltian commented:
> Is there a way for us to still support 3-field triples and generic targets?

I can't come up with a better solution because we don't know how many `-` could exist in the generic target id. With that, fixating the "known" part (aka triple) would be reasonable. Do you have any suggestions?

> In the past we worked to allow 3-field triples as bundler inputs. I would be surprised if adding a 4-field restriction didn't break something internally.

I don't know if there is any project using `clang-offload-bundler` directly. If there is, and we decide to enforce it, we will have to update them as well.

lamb-j commented:
https://github.com/search?q=org%3AROCm%20clang-offload-bundler&type=code

So at least a few direct users this may break things for. There are also users that call the OffloadBundler API instead of directly calling clang-offload-bundler that this may impact.

Could we assume "gfx * -generic" is a single Target ID, and check for/parse that separately before doing rsplit on "-"?

shiltian commented:
> Could we assume "gfx * -generic" is a single Target ID, and check for/parse that separately before doing rsplit on "-"?

I was thinking about it the other day but eventually gave up on it, because I want to keep this part as generic as possible instead of adding AMD special sauce here. I prefer not to do something like pattern matching stuff about `gfx`.

shiltian commented:
> So at least a few direct users this may break things for. There are also users that call the OffloadBundler API instead of directly calling clang-offload-bundler that this may impact.

For the API usage there will be no issue because we enforce it in the code already. The only part needs to be updated is those binary user.

jhuber6 commented:
Somewhere for the linker wrapper I just checked if the triple was recognized, you could probably just take strings after the `-` until it stops working.

yxsamliu commented:
> Somewhere for the linker wrapper I just checked if the triple was recognized, you could probably just take strings after the `-` until it stops working.

+1

It would be bad user experience to break existing app. It would be low risk to have env+cpu to be a valid cpu. So you could assume env exist first, if fails to parse remaining as cpu, then recoil to assume no env and parse the remaining all as cpu.

yxsamliu commented:
> First of all, I don't think it can fix the issue in a robust way. Second, `generic` is already a valid target/cpu/offload target.
> 
> Unless we do something like, if the last part is `generic`, we keep looking forward until we can construct a valid target. That has no difference than doing pattern matching, which is back to my previous point about AMD special sauce.
> 
> If we assert that the offload bundler is an AMD only thing (which TBH really looks like so), I'm fine with adding a bunch of more special sauce here.

offload-bundler is not an AMD only thing. At least HIPSPRV toolchain uses it, which is Intel GPU.

Still, I think it is possible to make it generic with minor assumption. Let's say you are now about to parsing the final part of the target ID string which may be either "env-cpu" or "cpu" without env. clang has a function getCanonicalProcessorName() which can check whether a string is a valid cpu name. Just pass the remaining string to it. If true, that means the remaining is a cpu, without env string. Otherwise, assuming there is an env string that contains no "-" and split it.

yxsamliu commented:
> > Still, I think it is possible to make it generic with minor assumption. Let's say you are now about to parsing the final part of the target ID string which may be either "env-cpu" or "cpu" without env.
> 
> This is not actually the issue. The issue is when the cpu is a generic target, such as `gfx10-3-generic`. By the current logic, the target id after split is `generic`, which is totally a valid one, and leave the rest with things like `hip-amd-amdhsa-amd-gfx10-3`.

That is probably due to this line https://github.com/llvm/llvm-project/blob/main/clang/lib/Driver/OffloadBundler.cpp#L88

It assumes there is no '-' in GPU name.

we could add a loop. If that line fails, we will split at the second '-' from right.

yxsamliu commented:
> > > Still, I think it is possible to make it generic with minor assumption. Let's say you are now about to parsing the final part of the target ID string which may be either "env-cpu" or "cpu" without env.
> > 
> > 
> > This is not actually the issue. The issue is when the cpu is a generic target, such as `gfx10-3-generic`. By the current logic, the target id after split is `generic`, which is totally a valid one, and leave the rest with things like `hip-amd-amdhsa-amd-gfx10-3`.
> 
> That is probably due to this line https://github.com/llvm/llvm-project/blob/main/clang/lib/Driver/OffloadBundler.cpp#L88
> 
> It assumes there is no '-' in GPU name.
> 
> we could add a loop. If that line fails, we will split at the second '-' from right.

Ok I get your point. since generic is a valid GPU name. it will stop there.

yxsamliu commented:
how about assuming the strict triple format first, this will make the generic GPU arch work with strict triple.

If the first assumption fails, then fall back to the legacy parsing, that is, assuming no '-' in GPU arch and split at the right most '-'. This way, the old target ID string with non-strict triple still works.

shiltian commented:
> how about assuming the strict triple format first, this will make the generic GPU arch work with strict triple.

That doesn't work if someone write a 3-tuple with a generic target because neither pass can handle that.

yxsamliu commented:
> > how about assuming the strict triple format first, this will make the generic GPU arch work with strict triple.
> 
> That doesn't work if someone write a 3-tuple with a generic target because neither path can handle that.

That's true. But your current approach does not work for that use case either, right? For the generic target, we request the 4-tuple. For the old target, we allow the 3-tuple to avoid regressions.

shiltian commented:
> For the generic target, we request the 4-tuple.

I don't think we can enforce that.

> For the old target, we allow the 3-tuple to avoid regressions.

I'm gonna submit PR to all the related projects (e.g. https://github.com/search?q=org%3AROCm%20clang-offload-bundler&type=code) to use 4-tuple.

lamb-j commented:
Could we possibly do a phase-out approach?

For COV5 and earlier we keep the current strategy, allowing both 3 and 4 field triples

For COV6 (including generics), we enforce the 4 field triples

This could help us not break things internally, and then as tools and APIs update to V6, they can also update to 4-field triples

shiltian commented:
Isn't offload bundler a target agnostic tool?

> This could help us not break things internally, and then as tools and APIs update to V6, they can also update to 4-field triples

PSDB has full green. I also created issues to all repos that use `clang-offload-bundler` binary directly. Most of them have responded and started using 4-tuple now. IMHO if we tell users, hey, if you do COV5, you can use 3- or 4-tuple, but if you do COV6, you have to use 4-tuple, it's gonna me way more confusing than directly enforcing them. In addition, internally we have switched to COV6 by default already for staging branch, where this PR is gonna go, and will be soon in mainline.

jhuber6 commented:
> Isn't offload bundler a target agnostic tool?

Would it break anything if we just assumed it was always `amdgcn-amd-amdhsa` for the triple? It's supposed to be generic, but right now only HIP uses it AFAICT.

shiltian committed:
[OffloadBundler] Rework the ctor of `OffloadTargetInfo` to support generic target

The current parsing of target string assumes to be in a form of
`kind-triple-targetid:feature`, such as
`hipv4-amdgcn-amd-amdhsa-gfx1030:+xnack`. Specifically, the target id does not
contain any `-`, which is not the case for generic target. Also, a generic
target may contain one or more `-`, such as `gfx10-3-generic` and
`gfx12-generic`. As a result, we can no longer depend on `rstrip` to get things
work right. This patch reworks the logic to parse the target string to make it
more robust, as well as supporting generic target.

yxsamliu commented on clang/lib/Driver/OffloadBundler.cpp:
```diff
@@ -84,31 +84,27 @@ OffloadTargetInfo::OffloadTargetInfo(const StringRef Target,
+  // <triple> := <arch>-<vendor>-<os>-<env>
+  SmallVector<StringRef, 6> Components;
+  Target.split(Components, '-', /*MaxSplit=*/5);
+  assert((Components.size() == 5 || Components.size() == 6) &&
```
We have users who executes clang-offload-bundler directly. What if they use `<arch>-<vendor>-<os>` as triple? Should we emit an error message instead of assert?

shiltian replied:
> We have users who executes clang-offload-bundler directly.

Not sure about people outside of ROCm stack. Every one in ROCm stack has already moved to 4-field triple.

This code is inside a constructor such that we can't emit an error here. If we really want to do that, we will need to duplicate the logic to somewhere else and emit an error there.

shiltian replied:
I added some error check mechanism but that is really limited. If someone still uses 3-field triple, they will get an error saying `<kind>-<arch>-<vendor>-<os>--` is not found.

We really can't tell whether a string is a valid data triple or not, given the flexibility of how data layout parsing is. However, it will error out later when it comes to comparison/

yxsamliu commented on clang/test/Driver/clang-offload-bundler.c:
```diff
@@ -520,32 +520,32 @@
 //
 // Create few code object bundles and archive them to create an input archive
-// RUN: clang-offload-bundler -type=o -targets=host-%itanium_abi_triple,openmp-amdgcn-amd-amdhsa-gfx906,openmp-amdgcn-amd-amdhsa--gfx908 -input=%t.o -input=%t.tgt1 -input=%t.tgt2 -output=%t.simple.bundle
+// RUN: clang-offload-bundler -type=o -targets=host-%itanium_abi_triple,openmp-amdgcn-amd-amdhsa--gfx906,openmp-amdgcn-amd-amdhsa--gfx908 -input=%t.o -input=%t.tgt1 -input=%t.tgt2 -output=%t.simple.bundle
```
need a test for using the old bundle entry ID hip-amdgcn-amd-amdhsa-gfx906 to bundle/unbundle

shiltian committed:
Update document

shiltian committed:
Add bundle id check such as it should not crash when given a malformat bundle id

shiltian committed:
Fix a test issue

yxsamliu reviewed (approved):
LGTM. Thanks

shiltian commented:
It seems like for some targets the compiler doesn't accept a 4-field triple. Why is that?

```
clang: error: version 'elf-unknown' in target triple 'hexagon-unknown-unknown-elf-unknown' is invalid
clang: error: version '-unknown' in target triple 'riscv64-unknown-linux-gnu-unknown' is invalid
```

shiltian commented:
Yeah, that needs to be fixed. The command line argument is `--target=hexagon-unknown-elf-unknown`, which is a valid target triple, but the error is `hexagon-unknown-unknown-elf-unknown`. That definitely exposes the issue in the compiler driver.

asb commented:
> It seems like some targets add "-unknown" in the compiler driver.
> 
> ```
> + /local/mnt/workspace/bots/hexagon-build-02/clang-hexagon-elf/stage1/bin/clang --target=hexagon-unknown-elf-unknown -fverbose-asm -g -S /local/mnt/workspace/bots/hexagon-build-02/clang-hexagon-elf/llvm/clang/test/CodeGenCXX/debug-info-member.cpp -o -
> clang: error: version 'elf-unknown' in target triple 'hexagon-unknown-unknown-elf-unknown' is invalid
> 
> + /home/buildbot-worker/bbroot/clang-riscv-rva20-2stage/stage2/bin/clang --target=riscv64-linux-gnu-unknown -g -S /home/buildbot-worker/bbroot/clang-riscv-rva20-2stage/llvm/clang/test/CodeGenCXX/debug-info-byval.cpp -o -
> clang: error: version '-unknown' in target triple 'riscv64-unknown-linux-gnu-unknown' is invalid
> FileCheck error: '<stdin>' is empty.
> FileCheck command line:  /home/buildbot-worker/bbroot/clang-riscv-rva20-2stage/stage2/bin/FileCheck /home/buildbot-worker/bbroot/clang-riscv-rva20-2stage/llvm/clang/test/CodeGenCXX/debug-info-byval.cpp
> ```

This is breaking multiple buildbots, so the correct fix is to revert and rework. I have erred in not spotting this breakage earlier as I incorrectly believed I'd receive an email as bot owner for a new failure. The correct path forwards should have been a revert https://llvm.org/docs/DeveloperPolicy.html#patch-reversion-policy though I know time has passed now.

@pratlucas did you get anywhere looking into your downstream breakage? CC @SundeepKushwaha on the hexagon side. I'll take a closer look tomorrow.

pratlucas commented:
Thanks for picking this up @asb ! Unfortunately I still couldn't find the time to investigate the root cause of the issue in our downstream runs. I'll keep you posted in case we find any more details on this.

shiltian commented:
@asb I did receive some emails initially but haven't gotten any lately. I assumed that issue had been fixed, as it seemed like the driver was just unconditionally adding `unknown`.
